packages/core/src/indexing-store/historical.ts

```diff
@@ -753,133 +793,169 @@ export const createHistoricalIndexingStore = ({
     ),
     async flush() {
       await queue.add(async () => {
-        const promises: Promise<any>[] = [];
+        let cacheSize = 0;
+        for (const c of cache.values()) cacheSize += c.size;
 
-        for (const [table, tableCache] of cache) {
-          const entries = Array.from(tableCache.values());
+        const flushIndex =
+          totalCacheOps -
+          cacheSize * (1 - common.options.indexingCacheFlushRatio);
+        const shouldDelete = cacheBytes > maxBytes;
+        if (shouldDelete) isDatabaseEmpty = false;
 
-          const batchSize = Math.round(
-            common.options.databaseMaxQueryParameters /
-              Object.keys(getTableColumns(table)).length,
-          );
+        const promises: Promise<void>[] = [];
 
-          const insertValues = entries
+        for (const [table, tableCache] of cache) {
+          const removableInsertValues = Array.from(tableCache.values())
```

Comment:
I'm curious if lines 808-851 could be done in a single pass over `tableCache.values()` (for loop + mutation vs. copies with map and filter).

Might not make a measurable difference though, not sure.